app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,267 @@
+        super(platform, platform.getGlobalResourceTable(),
+                platform.getUpgradeResourceTable(),
+                platform.getRecoveryTable());
+
+        app = CommCareApplication._().getCurrentApp();
+
+        tempUpgradeTable =
+                ResourceTable.RetrieveTable(app.getStorage(TEMP_UPGRADE_TABLE_KEY, Resource.class),
+                        new AndroidResourceInstallerFactory(app));
+
+        updateStats = UpdateStatPersistence.loadUpdateStats(app);
+        upgradeTable.setInstallStatsLogger(updateStats);
+        tempUpgradeTable.setInstallStatsLogger(updateStats);
+    }
+
+    /**
+     * Download the latest profile; if it is new, download and stage the entire
+     * update.
+     *
+     * @param profileRef Reference that resolves to the profile file used to
+     *                   seed the update
+     * @return UpdateStaged upon update download, UpToDate if no new update,
+     * otherwise an error status.
+     */
+    public AppInstallStatus checkAndPrepareUpgradeResources(String profileRef) {
+        synchronized (updateLock) {
+            this.profileRef = profileRef;
+            try {
+                instantiateLatestProfile();
+
+                if (isUpgradeTableStaged()) {
+                    return AppInstallStatus.UpdateStaged;
+                }
+
+                if (updateIsntNewer(getMasterProfile())) {
+                    Logger.log(AndroidLogger.TYPE_RESOURCES, "App Resources up to Date");
+                    upgradeTable.clear();
+                    return AppInstallStatus.UpToDate;
+                }
+
+                prepareUpgradeResources();
+            } catch (InstallCancelledException e) {
+                // The user cancelled the upgrade check process. The calling task
+                // should have caught and handled the cancellation
+                return AppInstallStatus.UnknownFailure;
+            } catch (LocalStorageUnavailableException e) {
+                ResourceInstallUtils.logInstallError(e,
+                        "Couldn't install file to local storage|");
+                return AppInstallStatus.NoLocalStorage;
+            } catch (UnfullfilledRequirementsException e) {
+                if (e.isDuplicateException()) {
+                    return AppInstallStatus.DuplicateApp;
+                } else {
+                    ResourceInstallUtils.logInstallError(e,
+                            "App resources are incompatible with this device|");
+                    return AppInstallStatus.IncompatibleReqs;
+                }
+            } catch (UnresolvedResourceException e) {
+                return ResourceInstallUtils.processUnresolvedResource(e);
+            }
+
+            return AppInstallStatus.UpdateStaged;
+        }
+    }
+
+    /**
+     * Load the latest profile into the upgrade table. Clears the upgrade table
+     * if it's partially populated with an out-of-date version.
+     */
+    private void instantiateLatestProfile()
```

Comment:
would be good to rename to something that makes it clear that it's for the upgrade table